### ISSUE TITLE
not yet implemented API endpoints in Foreman/Katello/Luna

### ISSUE THREAD

**Issue by evgeni** [theforeman/foreman-ansible-modules#1079]
##### SUMMARY
Ohai,

during some recent discussions with @ehelms, we wondered what the actual gap is between FAM and latest Foreman.

So I took a ride with `ast` and the apidocs we haven in our fixtures, and came up with:

```python
import ast
import glob
import json
import re

from plugins.module_utils import _apypie as apypie

# autodetection is not perfect
implemented_resources = {'capsule_content', 'content_uploads', 'content_view_filter_rules', 'filters', 'host_classes', 'hostgroup_classes', 'media', 'override_values', 'permissions', 'template_inputs', 'template_kinds', 'sync', 'puppetclasses', 'parameters', 'gpg_keys', 'content_view_components', 'subscriptions'}
ignored_resources = {'auth_sources', 'auth_source_externals', 'auth_source_internals', 'dashboard', 'home', 'interfaces', 'ping', 'plugins', 'hosts_bulk_actions', 'products_bulk_actions', 'repositories_bulk_actions', 'content_view_puppet_modules'}

inflector = apypie.Inflector()

for module_file_path in glob.iglob('plugins/modules/*.py'):
    with open(module_file_path) as module_file:
        module_ast = ast.parse(module_file.read())
        for entry in module_ast.body:
            if isinstance(entry, ast.ClassDef):
                class_name = re.sub(r'(?<=[a-z])[A-Z]|[A-Z](?=[^A-Z])', r'_\g<0>', entry.name).lower().strip('_')
                entity_name = '_'.join(class_name.split('_')[1:-1])
                resource_name = inflector.pluralize(entity_name)
                implemented_resources.add(resource_name)

for featureset in ['foreman', 'katello', 'luna']:
    print("\n{}:".format(featureset))
    existing_resources = set()
    with open('tests/fixtures/apidoc/{}.json'.format(featureset)) as apidoc_file:
        apidoc = json.load(apidoc_file)
        for resource_name, resource in apidoc['docs']['resources'].items():
            for method in resource['methods']:
                if any([api for api in method['apis'] if api['http_method'] != 'GET']):
                    existing_resources.add(resource_name)
    for missing in sorted(existing_resources - implemented_resources - ignored_resources):
        print("  - [ ] {}".format(missing))
    # let's pretend all we've seen is implemented now
    implemented_resources.update(existing_resources)

```

Which results in the following list (updated: 2021-06-17):

foreman:
  - [ ] autosign
  - [ ] config_reports
  - [ ] mail_notifications
  - [ ] personal_access_tokens
  - [ ] puppet_hosts
  - [ ] registration
  - [ ] report_templates
  - [ ] ssh_keys
  - [ ] table_preferences
  - [ ] template_combinations
  - [ ] trends

foreman_tasks (manually added):
  - [ ] foreman_tasks
  - [ ] recurring_logics

remote_execution (manually added):
  - [ ] foreign_input_sets
  - [ ] remote_execution_features

katello:
  - [ ] content_export_incrementals
  - [ ] content_exports
  - [ ] content_imports
  - [ ] host_errata
  - [ ] host_packages
  - [ ] host_subscriptions
  - [ ] host_tracer
  - [ ] package_groups
  - [ ] simple_content_access
  - [ ] upstream_subscriptions

luna:
  - [ ] ansible_inventories
  - [ ] ansible_override_values
  - [ ] ansible_roles
  - [ ] ansible_variables
  - [ ] arf_reports
  - [ ] configs
  - [ ] discovered_hosts
  - [x] discovery_rules
  - [ ] policies
  - [ ] templates

Pretty sure I didn't properly mark as implemented all available that weren't detected by the name, but it's a start.

We should talk which of these make sense to have, and create feature requests.

##### ISSUE TYPE
 - Feature Idea

**Comment by evgeni** [comment]
triage:
* re-run but don't look at controllers that only have GET routes.
* also look at foreman 2.3/ katello 3.18 etc

**Comment by evgeni** [comment]
script and result updated, the 2.3 apidoc didn't yield any changes in regard to stats/trends on itself, as those are still there, but documented as "deprecated" and "you need a plugin now"

**Comment by mdellweg** [comment]
I don't think we need to consider the bulk actions, because ansible can take care of the looping pattern here.

**Comment by ehelms** [comment]
Here are my thoughts on the Katello section.

katello:
  - [x] content_view_components: handled by content_view module [example](https://github.com/theforeman/foreman-ansible-modules/blob/develop/plugins/modules/content_view.py#L122)
  - [x] content_view_puppet_modules: goes away in Katello 4.0, suggest do not implement
  - [x] foreign_input_sets: templates related, not specific to Katello
  - [ ] host_errata: should implement
  - [ ] host_packages: should implement
  - [ ] host_subscriptions: should implement
  - [ ] host_tracer: should implement
  - [x] hosts_bulk_actions: Ansible loops should handle
  - [ ] package_groups: should implement, can create these, probably low use rate
  - [x] products_bulk_actions: Ansible loops should handle
  - [x] repositories_bulk_actions: Ansible loops
  - [x] subscriptions: handled by [manifest module](https://github.com/theforeman/foreman-ansible-modules/blob/develop/plugins/modules/subscription_manifest.py)
  - [ ] upstream_subscriptions: should implement, at least for SCA enablement

Let's add this section to the output of the script:
```
foreman_tasks:
  - [ ] foreman_tasks:
  - [ ] recurring_logics:
```

**Comment by evgeni** [comment]
thanks Eric, I've updated the Katello ones, but didn't add the tasks/rex split, as we don't have an apidoc that would have these features w/o katello

**Comment by Thulium-Drake** [comment]
#1431 addresses one of the items on the todo list :-)
